include/drivers/regulator.h

```diff
@@ -0,0 +1,163 @@
+#include <util.h>
+
+#define REGULATOR_OPS(dev) \
+	(&container_of((dev)->drv, struct regulator_driver, drv)->ops)
+
+enum {
+	REG_READABLE = BIT(0), /**< Regulator is readable via SCPI. */
+	REG_WRITABLE = BIT(1), /**< Regulator is writable via SCPI. */
+	REG_CRITICAL = BIT(4), /**< Regulator cannot be disabled. */
+};
+
+struct regulator_range {
+	const uint16_t start_raw;   /**< Smallest raw value in the range. */
+	const int16_t  start_value; /**< Cooked value at smallest raw value. */
+	const uint16_t step;        /**< Distance between adjacent values. */
+};
+
+struct regulator_info {
+	const char *const            name;      /**< Name exported to SCPI. */
+	const uint16_t               min_value; /**< Minimum allowed value. */
+	const uint16_t               max_value; /**< Maximum allowed value. */
+	const struct regulator_range ranges[2]; /**< Range descriptions. */
+	const uint8_t                flags;     /**< Generic class flags. */
+};
+
+struct regulator_driver_ops {
+	uint8_t                (*get_count)(struct device *dev);
+	struct regulator_info *(*get_info)(struct device *dev, uint8_t id);
+	int                    (*get_state)(struct device *dev, uint8_t id);
+	int                    (*read_raw)(struct device *dev, uint8_t id,
+	                                   uint32_t *raw);
+	int                    (*set_state)(struct device *dev, uint8_t id,
+	                                    bool enable);
+	int                    (*write_raw)(struct device *dev, uint8_t id,
+	                                    uint32_t raw);
+};
+
+struct regulator_driver {
+	const struct driver               drv;
+	const struct regulator_driver_ops ops;
+};
+
+/**
+ * Disable the output of a regulator. If the regulator does not have output
+ * on/off control, this function may have no effect on the hardware.
+ *
+ * This function may fail with:
+ *   EIO    There was a problem communicating with the hardware.
+ *   EPERM  The regulator is critical and cannot be disabled.
+ *
+ * @param dev   The device containing this regulator.
+ * @param id    The device-specific identifier for this regulator.
+ * @return      Zero on success; a defined error code on failure.
+ */
+int regulator_disable(struct device *dev, uint8_t id);
+
+/**
+ * Enable the output of a regulator. If the regulator does not have output
+ * on/off control, this function may have no effect on the hardware.
+ *
+ * This function may fail with:
+ *   EIO    There was a problem communicating with the hardware.
+ *
+ * @param dev   The device containing this regulator.
+ * @param id    The device-specific identifier for this regulator.
+ * @return      Zero on success; a defined error code on failure.
+ */
+static inline int
+regulator_enable(struct device *dev, uint8_t id)
+{
+	return REGULATOR_OPS(dev)->set_state(dev, id, true);
+}
+
+/**
+ * Get the number of regulators controlled by a device.
+ *
+ * This function has no defined errors.
+ *
+ * @param dev   The regulator device.
```

Comment:
Even though the regulator device is the device directly being manipulated by this function, this really isn't in the spirit of #119 (taking a different device than the other API functions). I would like to see this be the same device parameter as the others.

Reply:
Maybe the comment isn't clear. This *is* taking the same device as the other API functions. It just doesn't have an ID parameter because it's getting information about the regulator device in general, not one specific regulator channel. Suggestions on how to make that clear?

Reply:
Ah, I see. Perhaps "The regulator device controller"? Other ones could also say, "The regulator device controller containing the specified regulator".

Alternatively, "The regulator device containing the specified regulator" for the other descriptions would allow this one to make more sense.